ppoprf/src/ppoprf.rs

```diff
@@ -87,6 +148,143 @@ impl ProofDLEQ {
   }
 
+  fn verify_batch(
+    &self,
+    public_value: &RistrettoPoint,
+    p: &[RistrettoPoint], //P
+    q: &[RistrettoPoint], //Q
+  ) -> bool {
+    //(M, Z) = ComputeComposites(B, C, D)
+    let (m, z) = ProofDLEQ::compute_composites(None, public_value, p, q);
+
+    //t2 = ((s * A) + (c * B))
+    let t2 = (self.s * RISTRETTO_BASEPOINT_POINT) + (self.c * public_value);
+    //t3 = ((s * M) + (c * Z))
+    let t3 = (self.s * m) + (self.c * z);
+
+    /*challengeTranscript =
+    I2OSP(len(Bm), 2) || Bm ||
+    I2OSP(len(a0), 2) || a0 ||
+    I2OSP(len(a1), 2) || a1 ||
+    I2OSP(len(a2), 2) || a2 ||
+    I2OSP(len(a3), 2) || a3 ||
+    "Challenge"*/
+    let mut challenge_transcript = Vec::new();
+    challenge_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &public_value.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    challenge_transcript.extend_from_slice(public_value.compress().as_bytes()); //Bm
+    challenge_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &m.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    challenge_transcript.extend_from_slice(m.compress().as_bytes()); //a0 = m
+    challenge_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &z.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    challenge_transcript.extend_from_slice(z.compress().as_bytes()); //a1 = z
+    challenge_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &t2.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    challenge_transcript.extend_from_slice(t2.compress().as_bytes()); //a2 = t2
+    challenge_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &t3.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    challenge_transcript.extend_from_slice(t3.compress().as_bytes()); //a3 = t3
+
+    //c = G.HashToScalar(challengeTranscript)
+    let mut out = [0u8; 64];
+    strobe_hash(&challenge_transcript, "Challenge", &mut out);
+    let expected_c = RistrettoScalar::from_bytes_mod_order_wide(&out);
+
+    //verified = (expectedC == c)
+    return expected_c == self.c;
+  }
+
+  fn compute_composites(
+    key: Option<RistrettoScalar>,
+    b: &RistrettoPoint,
+    c: &[RistrettoPoint],
+    d: &[RistrettoPoint],
+  ) -> (RistrettoPoint, RistrettoPoint) {
+    if c.len() != d.len() {
+      panic!("C and D have a different number of elements!");
+    }
+
+    // We use the Partially-punctureable Oblivious Pseudo-Random Function
+    let context_string =
+      format!("{}{}{}{}", "PPOPRFv1-", 0x03, "-", "ristretto255-strobe");
+    //seedDST = "Seed-" || contextString
+    let seed_dst = format!("{}{}", "Seed-", context_string);
+
+    // seedTranscript = I2OSP(len(Bm), 2) || Bm || I2OSP(len(seedDST), 2) || seedDST
+    let mut seed_transcript = Vec::new();
+    seed_transcript.extend_from_slice(ProofDLEQ::i2osp(
+      &b.compress().as_bytes().len().to_be_bytes(),
+      2,
+    ));
+    seed_transcript.extend_from_slice(b.compress().as_bytes());
```

Comment:
This `b` is `b*m`, right?

Reply:
No, `b` is just `b`. In [draft-irtf-cfrg-voprf](https://cfrg.github.io/draft-irtf-cfrg-voprf/draft-irtf-cfrg-voprf.html#name-proof-verification) it is specified as `Bm = G.SerializeElement(B)`, where `SerializeElement` simply `Maps an Element A to a canonical byte array buf of fixed length Ne.`